### PR TITLE
[OCPCLOUD-1422] machine-api: azureprovider: add ultra disk support

### PR DESCRIPTION
**Pull request by damdo**
This PR implements the API changes described in https://github.com/openshift/enhancements/pull/1021

It adds options to allow the use of Ultra Disks (a.k.a. UltraSSDs) within Azure with Machine API.

Users will be able to use Ultra Disks in two ways:
- attach them as [Data Disks](https://docs.microsoft.com/en-us/azure/virtual-machines/managed-disks-overview#data-disk) to the Machine 
- attach and mount them via Persistent Volumes (PVs) to the Machine 

This enhancement is targeted for OCP `4.11`.

### CODE REVIEW

File: machine/v1beta1/types_azureprovider.go [hunk 1]
```diff
@@ -215,6 +223,48 @@ type OSDisk struct {
 	CachingType string `json:"cachingType,omitempty"`
 }
 
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+type DataDisk struct {
+	// NameSuffix is the suffix to be appended to the machine name to generate the disk name.
+	// Each disk name will be in format <machineName>_<nameSuffix>.
+	// NameSuffix name must start and finish with an alphanumeric character and can only contain letters, numbers, underscores, periods or hyphens.
+	// The overall disk name must not exceed 80 chars in length.
+	// +kubebuilder:validation:Pattern:=`^[a-zA-Z0-9](?:[\w\.-]*[a-zA-Z0-9])?$`
```
Comment (damdo) [inline machine/v1beta1/types_azureprovider.go:232]: Note: I had to change the regex a little from what was described in the enhancement (which I will update once this PR has been reviewed).
I had to use backticks rather than double quotes to wrap the regex, to make controller-gen happy (would've had to escape `\` otherwise). I've also pinned beginning and end of the string with `^` and `$`.

File: machine/v1beta1/types_azureprovider.go [hunk 2]
```diff
@@ -34,6 +34,9 @@ type AzureMachineProviderSpec struct {
 	Image Image `json:"image"`
 	// OSDisk represents the parameters for creating the OS disk.
 	OSDisk OSDisk `json:"osDisk"`
+	// DataDisk specifies the parameters that are used to add one or more data disks to the machine
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:37]: What gets placed on data disk versus an os disk?

File: machine/v1beta1/types_azureprovider.go [hunk 3]
```diff
@@ -83,6 +86,11 @@ type AzureMachineProviderSpec struct {
 	// SecurityProfile specifies the Security profile settings for a virtual machine.
 	// +optional
 	SecurityProfile *SecurityProfile `json:"securityProfile,omitempty"`
+	// UltraSSDCapability enables or disables Azure UltraSSD capability for a virtual machine.
+	// When omitted, the platform may enable the capability based on the configuration of data disks.
+	// +kubebuilder:validation:Enum:="Enabled";"Disabled"
+	// +optional
+	UltraSSDCapability AzureUltraSSDCapabilityState `json:"ultraSSDCapability,omitempty"`
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:93]: Is this related to the addition of datadisk?  The fact that the platform can set this almost suggests that certain datadisks should beinvalid unless this is enabled?

File: machine/v1beta1/types_azureprovider.go [hunk 4]
```diff
@@ -215,6 +223,48 @@ type OSDisk struct {
...
+	// NameSuffix name must start and finish with an alphanumeric character and can only contain letters, numbers, underscores, periods or hyphens.
+	// The overall disk name must not exceed 80 chars in length.
+	// +kubebuilder:validation:Pattern:=`^[a-zA-Z0-9](?:[\w\.-]*[a-zA-Z0-9])?$`
+	// +kubebuilder:validation:MaxLength:=78
+	// +kubebuilder:validation:Required
+	NameSuffix string `json:"nameSuffix"`
+	// DiskSizeGB is the size in GB to assign to the data disk.
+	// +kubebuilder:validation:Minimum=4
+	// +kubebuilder:validation:Required
+	DiskSizeGB int32 `json:"diskSizeGB"`
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:239]: Why an int32 instead of a resource.Quantity?

File: machine/v1beta1/types_azureprovider.go [hunk 5]
```diff
@@ -215,6 +223,48 @@ type OSDisk struct {
...
+	// ManagedDisk specifies the Managed Disk parameters for the data disk.
+	// +optional
+	ManagedDisk ManagedDiskParameters `json:"managedDisk,omitempty"`
+	// Lun Specifies the logical unit number of the data disk. This value is used to identify data disks within the VM and therefore must be unique for each data disk attached to a VM.
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:243]: Why do we let a user set this instead of choosing it ourselves?

File: machine/v1beta1/types_azureprovider.go [hunk 6]
```diff
@@ -215,6 +223,48 @@ type OSDisk struct {
...
+	// The value must be between 0 and 63.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=63
+	Lun *int32 `json:"lun,omitempty"`
+	// CachingType specifies the caching requirements.
+	// +optional
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:250]: what happens if not set?

File: machine/v1beta1/types_azureprovider.go [hunk 7]
```diff
@@ -229,7 +279,8 @@ type DiskSettings struct {
 
 // ManagedDiskParameters is the parameters of a managed disk.
 type ManagedDiskParameters struct {
-	// StorageAccountType is the storage account type to use. Possible values include "Standard_LRS" and "Premium_LRS".
+	// StorageAccountType is the storage account type to use.
+	// Possible values include "Standard_LRS", "Premium_LRS" and, only for Data Disks, "UltraSSD_LRS."
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:283]: Instead of having conditional validation, create a new type.  Starting with drift suggests that drift will continue.

Alternatively, let OSDisks use UltraSSD_LRS

File: machine/v1beta1/types_azureprovider.go (hunk 3 above)
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:93]: Why would a customer want to enable ultradisk capability on a per machineset basis?

Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:93]: If you want to use an UltraSSD as a PersistentVolume, then the capability needs to be enabled on the VM, as such we need a way for users to specify that they want it turned on when they don't have an UltraSSD directly attached.

We mention in the enhancement that if you set this to `Disabled` and specify an UltraSSD data disk, then that's invalid and we will reject the manifest. If you don't set a value and request an UltraSSD, we will enable it for you implicitly so that the disk is valid

Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:93]: See response to previous question, to allow PVs to use UltraSSD

File: machine/v1beta1/types_azureprovider.go (hunk 2 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:37]: OS Disk
_"Every virtual machine has one attached operating system disk. That OS disk has a pre-installed OS, which was selected when the VM was created. This disk contains the boot volume.
This disk has a maximum capacity of 4,095 GiB, however, many operating systems are partitioned with master boot record MBR by default. MBR limits the usable size to 2 TiB. If you need more than 2 TiB, create and attach [data disks](https://docs.microsoft.com/en-us/azure/virtual-machines/managed-disks-overview#data-disk) and use them for data storage"_

Data Disk
_"A data disk is a managed disk that's attached to a virtual machine to store application data, or other data you need to keep. Data disks are registered as SCSI drives and are labeled with a letter that you choose. Each data disk has a maximum capacity of 32,767 gibibytes"_

Also: _"Ultra Disks can't be used as OS Disks, they can only be created as empty Data Disks"_

Would you update the godoc comment to explain some if this in there?

File: machine/v1beta1/types_azureprovider.go (hunk 7 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:283]: Unfortunately we can't use Ultra Disks (`UltraSSD_LRS`) on OS Disks.
So would you split them into `OSDiskManagedDiskParameters` and `DataDiskManagedDiskParameters` for example?

File: machine/v1beta1/types_azureprovider.go (hunk 5 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:243]: This can be set both in the Azure Portal and in the upstream `kubernetes-sigs/cluster-api-provider-azure` implementation. So we thought we might want to favor compatibility with those and preserve the ability to manually specify them.

That said, if these `LUN`s are not set (and I would expect most users to do so), they will be be automatically chosen by the controller.

File: machine/v1beta1/types_azureprovider.go (hunk 6 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:250]: It won't be empty on the Machine struct underneath, which is then submitted to the cloud provider. This defaults it to the same default that would be chosen on the Azure GUI (Portal) if no options are specified for those. Which is `None`.

File: machine/v1beta1/types_azureprovider.go (hunk 4 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:239]: For a couple of reasons.
- This is the same of what is already set in OS Disk
- The [underlying Machine type provided](https://github.com/openshift/machine-api-provider-azure/blob/02da4328b3615ad84a2b808e203686fa4182d27d/pkg/cloud/azure/services/virtualmachines/virtualmachines.go#L263) by the cloud provider SDK accepts this as an `int32` on a field named `DiskSizeGB`
- This cannot be set at a finer grained level in the UI, they enforce it to be a Gigabyte integer
- In designing this, I tried to match the upstream API interface as close as possible to allow easier feature migration between MAPI and CAPI, and this is how they are implementing it

File: machine/v1beta1/types_azureprovider.go [hunk 8]
```diff
@@ -215,6 +223,48 @@ type OSDisk struct {
...
+	// +kubebuilder:validation:Pattern:=`^[a-zA-Z0-9](?:[\w\.-]*[a-zA-Z0-9])?$`
+	// +kubebuilder:validation:MaxLength:=78
+	// +kubebuilder:validation:Required
+	NameSuffix string `json:"nameSuffix"`
+	// DiskSizeGB is the size in GB to assign to the data disk.
+	// +kubebuilder:validation:Minimum=4
+	// +kubebuilder:validation:Required
+	DiskSizeGB int32 `json:"diskSizeGB"`
+	// ManagedDisk specifies the Managed Disk parameters for the data disk.
+	// +optional
```
[the remark this answers is not in the document]
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:241]: Similarly to what happens for the OSDisk, it is then passed down to the Machine struct underneath, which is then submitted to the cloud provider. This defaults ManagedDisk's fields to the same defaults that would be chosen on the Azure GUI (Portal) if no options were specified.
For Data Disk, it defaults to `StorageAccountType: PremiumSSD_LRS` and `DiskEncryptionSet: Default`.

File: machine/v1beta1/types_azureprovider.go (hunk 4 above)
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:239]: > This is the same of what is already set in OS Disk

sins of the past don't rule the future.

> The underlying Machine type provided by the cloud provider SDK accepts this as an int32 on a field named DiskSizeGB

This API follows kube conventions.  The choices of the underlying API don't matter.

> This cannot be set at a finer grained level in the UI, they enforce it to be a Gigabyte integer

This one is slightly more compelling, but I think the argument is stronger for, "the additional flexibility of resource.Quantity doesn't provide value to the user here"

> In designing this, I tried to match the upstream API interface as close as possible to allow easier feature migration between MAPI and CAPI, and this is how they are implementing it

This isn't hugely compelling.  A CAPI api mistake doesn't need to be reflected in our API.

File: machine/v1beta1/types_azureprovider.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:241]: Optional fields need to indicate in their godoc what happens when they are not included.  And whether it is intent based and subject to change to a guaranteed default.

File: machine/v1beta1/types_azureprovider.go (hunk 5 above)
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:243]: > So we thought we might want to favor compatibility with those and preserve the ability to manually specify them.

Is there a some user capability that requires this.  Exposure in other APIs doesn't mean it should be exposed in ours.

File: machine/v1beta1/types_azureprovider.go (hunk 6 above)
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:250]: As above, you need to indicate what the defaults are for *every* optional field and whether they are guaranteed forever or subject to change.

File: machine/v1beta1/types_azureprovider.go (hunk 7 above)
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:283]: > So would you split them into OSDiskManagedDiskParameters and DataDiskManagedDiskParameters for example?

Yes.  They have different values, validation, and documentation.

File: machine/v1beta1/types_azureprovider.go (hunk 4 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:239]: > the additional flexibility of resource.Quantity doesn't provide value to the user here

Yes that's probably the best way to put that :)

So, to come to a verdict, does it sound reasonable to keep the `int32` to you? Thanks

Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:239]: Yes, I find that reasoning good.

Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:239]: Yes, I find that reasoning good.

File: machine/v1beta1/types_azureprovider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:93]: Booleans are not allowed based on the OpenShift and Kubernetes API conventions. In particular, having booleans does not allow for a third state to determine if the user has not set the value. For example, here, we will imply from the fact that they have omitted the value but requested an UltraDisk for the OSDisk that it should be enabled.

In general, we aren't restricted by the upstream APIs of the cloud provider and should aim to provide something that is clear to users, I'm pretty happy that Enabled and Disabled are pretty obvious to end users

Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:93]: Update the comment to explain the PV case and to explain the failure mode if set to an impossible value.

Also, be sure to validate that someone isn't setting to disabled and trying to add an ultra disk.

File: machine/v1beta1/types_azureprovider.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:241]: I still don't see information about what happens when excluded.

File: machine/v1beta1/types_azureprovider.go (hunk 5 above)
Comment (deads2k) [reply machine/v1beta1/types_azureprovider.go:243]: > Although this doesn't rule out that there might be certain use cases where being able to specify a LUN it's a hard requirement.

Do we have any?  A "might be useful someday" isn't sufficient to provide an API field that could be added later if truly required.

File: machine/v1beta1/types_azureprovider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jstuever) [reply machine/v1beta1/types_azureprovider.go:93]: What I said is not excluded by the k8s convention... as `UltraSSDEnabled` can be `{'true','false',''}` and would be an obvious mapping to the targeted API parameter.

[the remark this answers is not in the document]
Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:93]: Would having a string value in the schema with `true`, `false` and `''` values work? I would have expected marshalling issues with that kind of schema? Something along the lines of `expected string, got bool`? Maybe it does work

File: machine/v1beta1/types_azureprovider.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:260]: Should be there now.

File: machine/v1beta1/types_azureprovider.go (hunk 5 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go]: Not really at the moment. Ok if you think we can add it later if truly required, I'm happy to drop it.

File: machine/v1beta1/types_azureprovider.go (hunk 6 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:265]: Done.

File: machine/v1beta1/types_azureprovider.go (hunk 7 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go]: Ok I kept the original `ManagedDiskParameters` to be used only for OS Disks, and created a new `DataDiskManagedDiskParameters`.

File: machine/v1beta1/types_azureprovider.go (hunk 2 above)
Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:37]: Additional information in the godoc for this field would be useful. I think telling the user why they might want a data disk or how they might use it, (I guess via a host volume?), would be useful

File: machine/v1beta1/types_azureprovider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:110]: Coming back to this, I'm still a firm believer that we should try to make our APIs readable and clear, and while having this different from the upstream Azure API is a source of possible confusion, I think the naming here is clear enough and, personally, more in-keeping with kube practices and naming. I don't believe this will be a source of confusion for users, and I think having the name this way makes it expandable in the future should the capability change at all

[the remark this answers is not in the document]
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:110]: I agree with Joel here, I'd rather not have booleans encoded as strings (can be misleading for the user and a source of error) just for the sake of keeping the same name present in the Azure API.

File: machine/v1beta1/types_azureprovider.go [hunk 9]
```diff
@@ -227,18 +292,33 @@ type DiskSettings struct {
 	EphemeralStorageLocation string `json:"ephemeralStorageLocation,omitempty"`
 }
 
-// ManagedDiskParameters is the parameters of a managed disk.
+// ManagedDiskParameters is the parameters of a OSDisk managed disk.
 type ManagedDiskParameters struct {
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:296]: You can rename the type as long you don't change any serialization.

File: machine/v1beta1/types_azureprovider.go [hunk 10]
```diff
@@ -227,18 +292,33 @@ type DiskSettings struct {
...
-	// StorageAccountType is the storage account type to use. Possible values include "Standard_LRS" and "Premium_LRS".
+	// StorageAccountType is the storage account type to use.
+	// Possible values include "Standard_LRS", "Premium_LRS".
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:298]: lack of verification worries me.  *after* this PR merges, let's open a PR that fixes the validation and we'll talk about what we can and cannot do.

File: machine/v1beta1/types_azureprovider.go [hunk 11]
```diff
@@ -227,18 +292,33 @@ type DiskSettings struct {
...
 	StorageAccountType string `json:"storageAccountType"`
 	// DiskEncryptionSet is the disk encryption set properties
 	// +optional
 	DiskEncryptionSet *DiskEncryptionSetParameters `json:"diskEncryptionSet,omitempty"`
 }
 
+// DataDiskManagedDiskParameters is the parameters of a DataDisk managed disk.
+type DataDiskManagedDiskParameters struct {
+	// StorageAccountType is the storage account type to use.
+	// Possible values include "Standard_LRS", "Premium_LRS" and "UltraSSD_LRS".
```
Comment (deads2k) [inline machine/v1beta1/types_azureprovider.go:308]: generated validation please.

File: machine/v1beta1/types_azureprovider.go (hunk 9 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go]: Ok, I was wondering that! Thanks :)
Renamed in https://github.com/openshift/api/pull/1119/commits/d2a93b66bad42fb7d316f51f802f5d22effd6ebc

File: machine/v1beta1/types_azureprovider.go (hunk 11 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:308]: Ok I've done so through an Enum and created a new type with the various const for it.

File: machine/v1beta1/types_azureprovider.go (hunk 10 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:298]: Yes, we could improve this with an `kubebuilder:validation:Enum` in a follow up PR.

File: machine/v1beta1/types_azureprovider.go (hunk 7 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go]: Now changed to `OSDiskManagedDiskParameters`.

File: machine/v1beta1/types_azureprovider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jstuever) [reply machine/v1beta1/types_azureprovider.go:110]: I was only trying to help by providing my insight based on my 20+ years of experience. As I have two cards blocked on this PR, I'll gladly step aside in favor of getting this merged sooner than later.

[the remark this answers is not in the document]
Comment (JoelSpeed) [reply machine/v1beta1/types_azureprovider.go:110]: @jstuever We definitely appreciate your input here and thank you for raising the topic for discussion. 

My team have been working very closely with the API review team over the last few months and in this particular case, are following advice from previous discussions about the inflexibility of boolean fields and how we don't have to be restricted by upstream API decisions, and should consider the UX for users of our systems to make our configuration APIs easy to understand, and in keeping with our conventions.

We discussed as a team the term `Enabled` and how that wouldn't fit in with Kubernetes naming conventions and how with that term in the name, it would mean the only possible values really are true and false. By moving the state of the capability into the value, we allow for future extensibility of the API field should new context arise (eg we may want an autodetect value, which is the current behaviour when omitted), and follow the patterns set out in prior APIs within OpenShift. 

As we are already close to merging this PR and have gone through several rounds of API review already, I'm going to ask @damdo to resolve this conversation, but if you have any further concerns, I'm happy to have a chat.

File: machine/v1beta1/types_azureprovider.go [hunk 12]
```diff
@@ -215,6 +240,56 @@ type OSDisk struct {
 	CachingType string `json:"cachingType,omitempty"`
 }
 
+// DataDisk specifies the parameters that are used to add one or more data disks to the machine.
+// A Data Disk is a managed disk that's attached to a virtual machine to store application data.
+// It differs from an OS Disk as it doesn't come with a pre-installed OS, and it cannot contain the boot volume.
+type DataDisk struct {
```
Comment (JoelSpeed) [inline machine/v1beta1/types_azureprovider.go:246]: I think we will want to add some notes to this to say that you need to add ignition configuration to a customer user data secret to perform the disk partition/format/mount operations, so it's clear users need to do that per disk. Can also add a note that the LUN will be needed to coordinate between the disk and its mount point on the host

Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:253]: Yes, good point. I was precisely doing that. Here it is: https://github.com/openshift/api/pull/1119/commits/b6963011c5947ae2bab1c22c6589efde95cc8c99

**Comment by JoelSpeed** [comment]
Thanks for the updates and additional research into why we need the LUN

/lgtm

**Comment by JoelSpeed** [comment]
This project is not a part of the no-ff process, as such QE, PX and Docs labels need to be added manually

/label px-approved
/label qe-approved
/label docs-approved

File: machine/v1beta1/types_azureprovider.go [hunk 13]
```diff
@@ -83,6 +86,28 @@ type AzureMachineProviderSpec struct {
...
+	// This may manifest in the Pod being stuck in `ContainerCreating` phase.
+	// This defaulting behaviour may be subject to change in future.
+	//
+	// When set to "Enabled", if the capability is available for the Machine based on the scope and limitations described above, the capability will be set on the Machine.
+	// This will thus allow UltraSSD both as Data Disks and Persistent Volumes.
+	// If set to "Enabled" when the capability can't be available due to scope and limitations, the Machine will go into "Failed" state.
+	//
+	// When set to "Disabled", UltraSSDs will not be allowed either as Data Disks nor as Persistent Volumes.
+	// In this case if any UltraSSDs are specified as Data Disks on a Machine, the Machine will go into a "Failed" state.
+	// If instead any UltraSSDs are backing the volumes (via Persistent Volumes) of any Pods scheduled on a Node which is backed by the Machine, the Pod may get stuck in `ContainerCreating` phase.
```
[the remark this answers is not in the document]
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:106]: Sounds, good! Thanks for your help and suggestions.

File: machine/v1beta1/types_azureprovider.go (hunk 10 above)
Comment (damdo) [reply machine/v1beta1/types_azureprovider.go:315]: Here is a draft PR where we can talk about what we can do with it: https://github.com/openshift/api/pull/1139

**Comment by damdo** [comment]
API design and further changes discussed here are now being ported back to the enhancement initially introduced here: https://github.com/openshift/enhancements/pull/1021 with https://github.com/openshift/enhancements/pull/1060.
